plugins/tekton/src/utils/pipeline-topology-utils.ts

```diff
@@ -0,0 +1,589 @@
+import { TaskRunKind } from '../types/taskRun';
+
+const createGenericNode: NodeCreatorSetup =
+  (type, width?, height?) => (name, data) => ({
+    id: name,
+    label: data?.label || name,
+    runAfterTasks: data?.runAfterTasks || [],
+    ...(data && { data }),
+    height: height ?? NODE_HEIGHT,
+    width: width ?? NODE_WIDTH,
+    type,
+  });
+
+const getMaxFinallyNode = (finallyTaskList: PipelineTaskWithStatus[]) => {
+  const sortedFinallyTaskList = [...finallyTaskList].sort(
+    (a, b) => b.name.length - a.name.length,
+  );
+  return sortedFinallyTaskList[0]?.name || '';
+};
+
+// Node variations
+export const createTaskNode: NodeCreator<TaskNodeModelData> = createGenericNode(
+  NodeType.TASK_NODE,
+);
+export const createSpacerNode: NodeCreator<SpacerNodeModelData> =
+  createGenericNode(NodeType.SPACER_NODE, 0);
+export const createTaskListNode: NodeCreator<TaskListNodeModelData> =
+  createGenericNode(NodeType.TASK_LIST_NODE);
+export const createInvalidTaskListNode: NodeCreator<TaskListNodeModelData> =
+  createGenericNode(NodeType.INVALID_TASK_LIST_NODE);
+export const createBuilderNode: NodeCreator<BuilderNodeModelData> =
+  createGenericNode(NodeType.BUILDER_NODE);
+
+export const createFinallyNode = (
+  height: number,
+): NodeCreator<FinallyNodeModel> =>
+  createGenericNode(
+    NodeType.FINALLY_NODE,
+    NODE_WIDTH + WHEN_EXPRESSION_SPACING + FINALLY_NODE_PADDING * 2,
+    height,
+  );
+
+export const createLoadingNode: NodeCreator<LoadingNodeModel> =
+  createGenericNode(NodeType.LOADING_NODE);
+
+const createPipelineTaskNode = (
+  type: NodeType,
+  data: PipelineRunAfterNodeModelData,
+) => createGenericNode(type, data.width, data.height)(data.id ?? '', data);
+
+export const getNodeCreator = (
+  type: NodeType,
+): NodeCreator<PipelineRunAfterNodeModelData> => {
+  switch (type) {
+    case NodeType.TASK_LIST_NODE:
+      return createTaskListNode as NodeCreator<PipelineRunAfterNodeModelData>;
+    case NodeType.BUILDER_NODE:
+      return createBuilderNode as NodeCreator<PipelineRunAfterNodeModelData>;
+    case NodeType.SPACER_NODE:
+      return createSpacerNode;
+    case NodeType.LOADING_NODE:
+      return createLoadingNode as NodeCreator<PipelineRunAfterNodeModelData>;
+    case NodeType.INVALID_TASK_LIST_NODE:
+      return createInvalidTaskListNode as NodeCreator<PipelineRunAfterNodeModelData>;
+    case NodeType.TASK_NODE:
+    default:
+      return createTaskNode;
+  }
+};
+
+export const handleParallelToParallelNodes = (
+  nodes: PipelineMixedNodeModel[],
+): PipelineMixedNodeModel[] => {
+  type ParallelNodeReference = {
+    node: PipelineTaskNodeModel;
+    runAfter: string[];
+    atIndex: number;
+  };
+  type ParallelNodeMap = {
+    [id: string]: ParallelNodeReference[];
+  };
+
+  // Collect only multiple run-afters
+  const multipleRunBeforeMap: ParallelNodeMap = nodes.reduce(
+    (acc, node, idx) => {
+      const {
+        data: {
+          task: { runAfter },
```

Comment:
will `task` always be there? if not good to add check

Reply:
yes, each node will have task info and each task might not have 'runAfter' task(s) which is why there is a check on runAfter